File: Data Structures & Algorithms/reconstruct-flight-path/submission-8.py
"""
1. sort the tickets so we can access cities by lexi-order
2. pop the top node from the stack and add its neighbors to the stack
3. if a node have no neighbors, put it to itin
4. repeat 1 - 3 until empty stack
"""

class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        tickets.sort(reverse = True)
        # tickets = sorted(tickets, reverse = True)
        graph = defaultdict(list)
        for src, dst in tickets:
            graph[src].append(dst)
        stack = ["JFK"]
        itin = []
        while stack:
            while graph[stack[-1]]:
                stack.append(graph[stack[-1]].pop())
            itin.append(stack.pop())
        return itin[::-1]
            

File: Data Structures & Algorithms/evaluate-reverse-polish-notation/submission-8.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> lifo;
        for (string token : tokens) {
            if (token == "+") {
                int n1 = lifo.top();
                lifo.pop();
                int n2 = lifo.top();
                lifo.pop();
                lifo.push(n1 + n2);
            } else if (token == "*") {
                int n1 = lifo.top();
                lifo.pop();
                int n2 = lifo.top();
                lifo.pop();
                lifo.push(n1 * n2);
            } else if (token == "-") {
                int n1 = lifo.top();
                lifo.pop();
                int n2 = lifo.top();
                lifo.pop();
                lifo.push(n2 - n1);
            } else if (token == "/") {
                int n1 = lifo.top();
                lifo.pop();
                int n2 = lifo.top();
                lifo.pop();
                lifo.push(n2 / n1);
            } else {
                lifo.push(stoi(token));
            }
        }
        return lifo.top();
    }
};
